Add BigQuery Loader V2 (#877)
snowplow/documentation: docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/upgrade-guides/2-0-0-upgrade-guide/index.md

---
title: "2.0.0 upgrade guide"
sidebar_position: -20
---

## Configuration

BigQuery Loader 2.0.0 brings changes to the loading setup. It is no longer neccessary to configure and deploy three independent applications (Loader, Repeater and Mutator in [1.X](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/previous-versions/bigquery-loader-1.x/index.md)) in order to load your data to BigQuery.

> Starting from 2.0.0 only one appliction is needed, which naturally introduces some breaking changes to the configuration file structure.

Reviewer: ```suggestion
Starting from 2.0.0, only one appliction is needed, which naturally introduces some breaking changes to the configuration file structure.
```

See the [configuration reference](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/configuration-reference/index.md) for all possible configuration parameters
and the minimal [configuration samples](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/v2/config) for each of supported cloud environments.

## Infrastructure

> Apart from Repeater and Mutator, there are also other infrastructure components that have become obsolete:

Reviewer: ```suggestion
Apart from Repeater and Mutator, other infrastructure components have become obsolete:
```

* The `types` PubSub topic connecting Loader and Mutator.
* The `failedInserts` PubSub topic connecting Loader and Repeater.
* The `deadLetter` GCS bucket used by Repeater to store data that repeatedly failed to be inserted into BigQuery.

## Events table format

Starting from 2.0.0, BigQuery Loader changes its output column naming strategy. For example, for [ad_click event](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow.media/ad_click_event/jsonschema/1-0-0):

> * Before an upgrade, new column would be named `unstruct_event_com_snowplowanalytics_snowplow_media_ad_click_event_1-0-0`.

Reviewer: ```suggestion
* Before an upgrade, the corresponding column would be named `unstruct_event_com_snowplowanalytics_snowplow_media_ad_click_event_1-0-0`.
```

* After an upgrade, new column will be named `unstruct_event_com_snowplowanalytics_snowplow_media_ad_click_event_1`.

> This means all new self-describing events and entities will be loaded to 'major version' - oriented columns, whereas old 'full version' - oriented columns remain unchanged, are no longer used by the new loader, and have no direct effect on loading. If neccessary, old columns need to be consolidated separately. If you are [modeling your data with dbt](/docs/modeling-your-data/modeling-your-data-with-dbt/index.md), you can use [this macro](https://github.com/snowplow/dbt-snowplow-utils#combine_column_versions-source) to aggregate the data across multiple columns.

Reviewer: ```suggestion
All self-describing events and entities will be loaded to new 'major version' - oriented columns. Old 'full version' - oriented columns would remain unchanged, but no new data would be loaded into them. (The 2.0.0 loader just ignores these columns.)

The new column naming scheme has several advantages:
* Fewer columns created (BigQuery has a limit on the total number of columns)
* No need to update data models (or use complex macros) every time a new minor version of a schema is created

The catch is that you have to follow the rules of schema evolution more strictly to ensure data from different schema versions can fit in the same column — see below.

:::tip Consolidating old and new columns

If you are using [Snowplow dbt models](/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-models/index.md), they will automatically consolidate the data between `_1_0_0` and `_1` style columns, because they look at the major version prefix (e.g. `_1`), which is common to both.

If you are not using Snowplow dbt models but still use dbt, you can employ [this macro](https://github.com/snowplow/dbt-snowplow-utils#combine_column_versions-source) to manually aggregate the data across old and new columns.

:::
```

**Non-breaking**: The schema version can be changed in a minor way (`1-2-3` → `1-3-0` or `1-2-3` → `1-2-4`). Data is stored in the same database column.

### Without recovery columns 

Loader tries to format the incoming data according to the latest version of the schema it saw (for a given major version, e.g. `1-*-*`). For example, if a batch contains events with schema versions `1-0-0`, `1-0-1` and `1-0-2`, the loader derives the output schema based on version `1-0-2`. Then the loader instructs BigQuery to adjust the database column and load the data.

This logic relies on two assumptions:

1. **Old events compatible with new schemas.** Events with older schema versions, e.g. `1-0-0` and `1-0-1`, have to be valid against the newer ones, e.g. `1-0-2`. Those that are valid will result in failed events.

2. **Old columns compatible with new schemas.** The corresponding BigQuery columns have to be migrated correctly from one version to another. Changes, such as altering the type of a field from `integer` to `string`, would fail. Loading would break with SQL errors and the whole batch would be stuck and hard to recover.

> These assumptions are not always clear to the users, making the process error-prone.

Reviewer: This does not apply to BigQuery Loader. I guess it needs to be replaced with an explanation of why it was not relevant for Version 1.

### With recovery columns

> First, we support schema evolution that’s not strictly backwards compatible (although we still recommend against it since it can confuse downstream consumers of the data). This is done by _merging_ multiple schemas so that both old and new events can coexist. For example, suppose we have these two schemas:

Reviewer: Does this logic apply here? Or only to Databricks?

Reviewer: Either way I think it’s not relevant to BigQuery as _everything_ was allowed before, including this.

```json
{
   // 1-0-0
   "properties": {
      "a": {"type": "integer"}
   }
}
```

```json
{
   // 1-0-1
   "properties": {
      "b": {"type": "integer"}
   }
}
```

These would be merged into the following:
```json
{
   // merged
   "properties": {
      "a": {"type": "integer"},
      "b": {"type": "integer"}
   }
}
```

> Second, the loader does not fail when it can’t modify the database column to store both old and new events. (As a reminder, an example would be changing the type of a field from `integer` to `string`.) Instead, it creates a _temporary_ column for the new data as an exception. The users can then run SQL statements to resolve this situation as they see fit. For instance, consider these two schemas:

Reviewer: This needs rewording — old loader would not fail either

```json
{
   // 1-0-0
   "properties": {
      "a": {"type": "integer"}
   }
}
```

```json
{
   // 1-0-1
   "properties": {
      "a": {"type": "string"}
   }
}
```

Because `1-0-1` events cannot be loaded into the same column with `1-0-0`, the data would be put in a separate column, e.g. `contexts_com_snowplowanalytics_ad_click_1_0_1_recovered_9999999`, where:
  - `1_0_1` is the version of the offending schema;
  - `9999999` is a hash code unique to the schema (i.e. it will change if the schema is overwritten with a different one).

If you create a new schema `1-0-2` that reverts the offending changes and is again compatible with `1-0-0`, the data for events with that schema will be written to the original column as expected.

> ### Notes
> 
> - If events with incorrectly evolved schemas do not arrive, then the recovery column would not be created.
> - It is still possible to break loading by overwriting version `1-0-0` of the schema. Please, avoid doing that.

Reviewer: ```suggestion
:::note

If events with incorrectly evolved schemas never arrive, then the recovery column would not be created.

:::
```

> You can read more about schema evolution and how recovery columns work [here](/docs/storing-querying/schemas-in-warehouse/?warehouse=bigquery#versioning).

Reviewer: ```suggestion
You can read more about schema evolution and how recovery columns work [here](/docs/storing-querying/schemas-in-warehouse/index.md?warehouse=bigquery#versioning).
```

---
title: "2.0.0 upgrade guide"
sidebar_position: -20
---

## Configuration

> BigQuery Loader 2.0.0 brings changes to the loading setup. It is no longer neccessary to configure and deploy three (in [1.X](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/previous-versions/bigquery-loader-1.x/index.md) it was loader, repeater and mutator) independent applications in order to load your data to BigQuery.

Reviewer 2: Beyond just configuration... we also make obsolete some of the cloud infrastructure.  E.g. the pubsub topic that connected the 3 components.  And the GCS bucket where the repeater loaded failed inserts.

I think we can re-write this paragraph to improve readability.  But please add that extra info first, and then we can improve readability after we've got the content.

> BigQuery Loader 2.0.0 brings changes to the loading setup. It is no longer neccessary to configure and deploy three (in [1.X](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/previous-versions/bigquery-loader-1.x/index.md) it was loader, repeater and mutator) independent applications in order to load your data to BigQuery.

Reviewer: ```suggestion
BigQuery Loader 2.0.0 brings changes to the loading setup. It is no longer neccessary to configure and deploy three independent applications (Loader, Repeator and Mutator in [1.X](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/previous-versions/bigquery-loader-1.x/index.md)) in order to load your data to BigQuery.
```

Starting from 2.0.0 only one appliction is needed, which naturally introduces some breaking changes to the configuration file structure.

See the [configuration reference](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/configuration-reference/index.md) for all possible configuration parameters
and the minimal [configuration samples](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/v2/config) for each of supported cloud environments.

## Events table format

> Starting from 2.0.0, BigQuery Loader no longer uses full schema version in column names for self-describing events and entities in `events` table. It uses only major schema version in the column name instead.

Reviewer 2: I suggest drop in an example here.  Pick a commonly used schema, and say what the column is called before and after the upgrade.  Will be much quicker for the user to grasp, rather than explaining it with words.

After that, I think you can condense the first two paragraphs into fewer words.

> This means all new self-describing events and entities will be loaded to 'major version' - oriented columns, whereas old 'full version' - oriented columns remain unchanged, are no longer used by the new loader, and have no direct effect on loading. If neccessary, old columns need to be handled/consolidated separately. //TODO how?

Reviewer 2: The point about consolidating old columns is important.  You can ask our analytics engineers for help with phrasing this advice.

I _think_ the advice is..... if you use Snowplow's standard DBT models then don't worry it will all work ok automatically.  If you don't use Snowplow's DBT models then go figure it out for yourselves.

> Before 2.0.0, breaking changes introduced for the same schema family (shemas using the same major version) had no impact on your `events` table. Starting from 2.0.0, loader tries to merge all changes from the same schema family and load to the single column (with major version suffix). In case of breaking changes, loader creates recovery columns to try to load all your data, even the one referencing 'broken' schemas. You can read more about schema evolution and how recovery columns work [here](/docs/storing-querying/schemas-in-warehouse/?warehouse=bigquery#versioning).

Reviewer: I think this needs a concrete example, similar to what we had in Discourse announcements for recovery columns/tables in Databricks/Redshift

Starting from 2.0.0, only one appliction is needed, which naturally introduces some breaking changes to the configuration file structure.

See the [configuration reference](/docs/pipeline-components-and-applications/loaders-storage-targets/bigquery-loader/configuration-reference/index.md) for all possible configuration parameters
and the minimal [configuration samples](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/v2/config) for each of supported cloud environments.

## Infrastructure

Apart from Repeater and Mutator, other infrastructure components have become obsolete:
* The `types` PubSub topic connecting Loader and Mutator.
* The `failedInserts` PubSub topic connecting Loader and Repeater.
* The `deadLetter` GCS bucket used by Repeater to store data that repeatedly failed to be inserted into BigQuery.

## Events table format

Starting from 2.0.0, BigQuery Loader changes its output column naming strategy. For example, for [ad_click event](https://github.com/snowplow/iglu-central/blob/master/schemas/com.snowplowanalytics.snowplow.media/ad_click_event/jsonschema/1-0-0):

> * Before an upgrade, the corresponding column would be named `unstruct_event_com_snowplowanalytics_snowplow_media_ad_click_event_1-0-0`.

Reviewer: ```suggestion
* Before an upgrade, the corresponding column would be named `unstruct_event_com_snowplowanalytics_snowplow_media_ad_click_event_1_0_0`.
```
Right?

The new column naming scheme has several advantages:
* Fewer columns created (BigQuery has a limit on the total number of columns)
* No need to update data models (or use complex macros) every time a new minor version of a schema is created

The catch is that you have to follow the rules of schema evolution more strictly to ensure data from different schema versions can fit in the same column — see below.

:::tip Consolidating old and new columns

If you are using [Snowplow dbt models](/docs/modeling-your-data/modeling-your-data-with-dbt/dbt-models/index.md), they will automatically consolidate the data between `_1_0_0` and `_1` style columns, because they look at the major version prefix (e.g. `_1`), which is common to both.

If you are not using Snowplow dbt models but still use dbt, you can employ [this macro](https://github.com/snowplow/dbt-snowplow-utils#combine_column_versions-source) to manually aggregate the data across old and new columns.

:::

## Recovery columns

### What is schema evolution?

> One of Snowplow’s key features is the ability to [define custom schemas and validate events against them](https://docs.snowplow.io/docs/understanding-tracking-design/understanding-schemas-and-validation/). Over time, users often evolve the schemas, e.g. by adding new fields or changing existing fields. To accommodate these changes, BigQuery Loader 2.0.0 automatically adjusts the database tables in the warehouse accordingly.

Reviewer: ```suggestion
One of Snowplow’s key features is the ability to [define custom schemas and validate events against them](/docs/understanding-tracking-design/understanding-schemas-and-validation/index.md). Over time, users often evolve the schemas, e.g. by adding new fields or changing existing fields. To accommodate these changes, BigQuery Loader 2.0.0 automatically adjusts the database tables in the warehouse accordingly.
```

There are two main types of schema changes:

**Breaking**: The schema version has to be changed in a major way (`1-2-3` → `2-0-0`). As of BigQuery Loader 2.0.0, each major schema version has its own column (`..._1`, `..._2`, etc, for example: `contexts_com.snowplowanalytics_ad_click_1`).

**Non-breaking**: The schema version can be changed in a minor way (`1-2-3` → `1-3-0` or `1-2-3` → `1-2-4`). Data is stored in the same database column.

> Loader tries to format the incoming data according to the latest version of the schema it saw (for a given major version, e.g. `1-*-*`). For example, if a batch contains events with schema versions `1-0-0`, `1-0-1` and `1-0-2`, the loader derives the output schema based on version `1-0-2`. Then the loader instructs BigQuery to adjust the database column and load the data.

Reviewer: I think this does not apply exactly like this because there are no batches.

Author: There are batches, but more kind of... micro-batches? And within those micro-batches loader derives output schema.

### Recovering from invalid schema evolution 

Let's consider these two schemas as an example of breaking schema evolution (changing the type of a field from `integer` to `string`) using the same major version (`1-0-0` and `1-0-1`):

```json
{
   // 1-0-0
   "properties": {
      "a": {"type": "integer"}
   }
}
```

```json
{
   // 1-0-1
   "properties": {
      "a": {"type": "string"}
   }
}
```

> Because `1-0-1` events cannot be loaded into the same column with `1-0-0`, the data would be put in a separate column, e.g. `contexts_com_snowplowanalytics_ad_click_1_0_1_recovered_9999999`, where:

Reviewer: ```suggestion
With BigQuery Loader 1.x, data for each version would go to its own column — no issue. With BigQuery Loader 2.x, there is only one column. But strings and integers can’t coexist!

To avoid crashing or losing data, BigQuery Loader 2.0.0 proceeds by creating a new column for the data with schema `1-0-1`, e.g. `contexts_com_snowplowanalytics_ad_click_1_0_1_recovered_9999999`, where:
```

  - `1_0_1` is the version of the offending schema;
  - `9999999` is a hash code unique to the schema (i.e. it will change if the schema is overwritten with a different one).

If you create a new schema `1-0-2` that reverts the offending changes and is again compatible with `1-0-0`, the data for events with that schema will be written to the original column as expected.

> 

Reviewer: ```suggestion
:::tip
You might find that some of your schemas were evolved incorrectly in the past, which results in the creation of these “recovery” columns after the upgrade. To address this for a given schema, create a new _minor_ schema version that reverts the breaking changes introduced in previous versions. (Or, if you want to keep the breaking change, create a new _major_ schema version.) You can set it to [supersede](/docs/understanding-tracking-design/versioning-your-data-structures/amending/index.md#marking-the-schema-as-superseded) the previous version(s), so that events are automatically validated against the new schema.
:::
```

Do we also need instructions similar to these? https://docs.snowplow.io/docs/pipeline-components-and-applications/loaders-storage-targets/snowplow-rdb-loader/upgrade-guides/6-0-0-upgrade-guide/#identifying-schemas-that-need-patching @istreeter

Reviewer 2: That section "Identifying schemas that need patching" was written long ago before we had the supercededBy feature.  I would think now we have supercededBy, we shouldn't recommend to anyone to patch schemas.

Instead, the user can publish the fixed schema as a new major version, and configure the old schemas to automatically upgrade to the new major version.  This is consistent with our belief that schemas are immutable.  And it means we don't need to write a new igluctl feature.

Reviewer: My doubt was not specifically about patching but whether running igluctl would be useful to identify issues with schemas... Because I think it says what the incompatibilities are? With just a recovery column, how will the users know what to fix?

Reviewer 2: igluctl has an existing sub-command `verify-parquet`

```
igluctl static verify-parquet <schema_dir>
```

It prints out warnings like

```
'com.test/test/jsonschema/1-0-1'. Changes: Incompatible type change String to Double at /id"
```

Unfortunately, it only works if the schemas are written into files on the local file system.  It does not connect to remote Iglus.

We've not really planned what happens when existing users migrate to the new loader.  And that is semi-deliberate: it is a lot of work to gather all the requirements and plan for all the things that could go wrong.  So I've been delaying it until there is time to work on it properly.

Author: @stanch @istreeter as BQ Loader v2 relies on the same schema compatibility model/rules as parquet/Databricks/lake loader, we could use `verify-parquet` igluctl command and recommend it to users to detect incompatibilities. But this `parquet` in the name is confusing! Should we also do some renaming in igluctl then to make it more generic? And then also adjust [docs](https://deploy-preview-877--snowplow-docs.netlify.app/docs/pipeline-components-and-applications/iglu/igluctl-2/#igluctl)
